safe/new_impact_function/impact_function.py

```diff
@@ -0,0 +1,766 @@
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+"""
+from PyQt4.QtCore import QVariant
+
+from qgis.core import (
+    QgsMapLayer,
+    QgsCoordinateReferenceSystem,
+    QgsRectangle,
+    QgsVectorLayer,
+    QgsGeometry,
+    QgsFeature,
+    QgsField,
+    QgsDistanceArea,
+)
+
+import logging
+
+from definitionsv4.post_processors import post_processors
+from safe.defaults import get_defaults
+from safe.common.exceptions import (
+    InvalidExtentError, InvalidLayerError, NoKeywordsFoundError)
+from safe.utilities.i18n import tr
+from safe.utilities.keyword_io import KeywordIO
+
+LOGGER = logging.getLogger('InaSAFE')
+
+
+__author__ = 'ismailsunni'
+__project_name__ = 'inasafe-dev'
+__filename__ = 'impact_function'
+__date__ = '8/16/16'
+__copyright__ = 'imajimatika@gmail.com'
+
+
+def evaluate_formula(formula, variables):
+    """Very simple formula evaluator. Beware the security.
+    :param formula: A simple formula.
+    :type formula: str
+
+    :param variables: A collection of variable (key and value).
+    :type variables: dict
+
+    :returns: The result of the formula execution.
+    :rtype: float, int
+    """
+    for key, value in variables.items():
+        formula = formula.replace(key, str(value))
+    return eval(formula)
+
+
+class ImpactFunction(object):
+    """Impact Function."""
+
+    def __init__(self):
+        self._hazard = None
+
+        self._exposure = None
+
+        self._aggregation = None
+
+        # Requested extent to use
+        self._requested_extent = None
+        # Requested extent's CRS
+        self._requested_extent_crs = QgsCoordinateReferenceSystem('EPSG:4326')
+        # The current viewport extent of the map canvas
+        self._viewport_extent = None
+        # Actual extent to use - Read Only
+        self._actual_extent = None
+        # Actual extent's CRS - Read Only
+        self._actual_extent_crs = QgsCoordinateReferenceSystem('EPSG:4326')
+        # set this to a gui call back / web callback etc as needed.
+        self._callback = self.console_progress_callback
+
+        self.algorithm = None
+        self.impact_layer = None
+
+        self._name = None  # e.g. Flood Raster on Building Polygon
+        self._title = None  # be affected
+
+        self.state = {}
+        self.reset_state()
+
+    @property
+    def hazard(self):
+        """Property for the hazard layer to be used for the analysis.
+
+        :returns: A map layer.
+        :rtype: QgsMapLayer
+        """
+        return self._hazard
+
+    @hazard.setter
+    def hazard(self, layer):
+        """Setter for hazard layer property.
+
+        :param layer: Hazard layer to be used for the analysis.
+        :type layer: QgsMapLayer
+        """
+        self._hazard = layer
+        try:
+            self._hazard.keyword = KeywordIO().read_keywords(layer)
```

Comment:
I started to use `keywords`, not `keyword`. It seems more logical as we have many keyword ;-)